README.md
R package processR  <img src="inst/figures/imgfile.png" align="right" height="120" width="103.6"/>
=========================================================
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/processR)](https://cran.r-project.org/package=processR)


The `processR` package aims to be a user-friendly way to perform moderation, mediation, moderated mediation and moderated moderation in R. This package is inspired from famous PROCESS macro for SPSS and SAS created by Andrew Hayes. 

**processR is under the GPL-3 license. For a commercial license, please
[contact me](mailto: cardiomoon@gmail.com).**

## PROCESS macro and R package `processR`

Andrew F. Hayes was not involved in the development of this R package or application and cannot attest to the quality of the computations implemented in the code you are using. Use at your own risk.

## Installation

You can install the `processR` package from github.


```r
if(!require(devtools)) install.packages("devtools")
devtools::install_github("cardiomoon/processR")
```
 
## What does this package cover ? 

The `processR` package covers moderation, mediation, moderated mediation and moderated moderation with R. Supporting models are as follows.


```r
library(processR)
sort(pmacro$no)
```

```
 [1]  0.0  1.0  2.0  3.0  4.0  4.2  5.0  6.0  6.3  6.4  7.0  8.0  9.0 10.0
[15] 11.0 12.0 13.0 14.0 15.0 16.0 17.0 18.0 19.0 20.0 21.0 22.0 23.0 24.0
[29] 28.0 29.0 30.0 31.0 35.0 36.0 40.0 41.0 45.0 49.0 50.0 58.0 59.0 60.0
[43] 61.0 62.0 63.0 64.0 65.0 66.0 67.0 74.0 75.0 76.0
```

Currently, 52 models are supported.

## Example: Moderated Mediation (PROCESS macro model 8)

I will explain functions of processR package by a example.

## Concept Diagram and Statistical Diagram

You can draw concept diagram and statistical diagram easily. For example, you can draw the concept diagram for PROCESS macro model 8.


```r
pmacroModel(8)
```

<img src="https://github.com/cardiomoon/processR/blob/master/figure/fig1.png?raw=true"  style="display: block; margin: auto;" />


You can draw statistical diagram of this model.


```r
statisticalDiagram(8)
```

<img src="https://github.com/cardiomoon/processR/blob/master/figure/fig2.png?raw=true"  style="display: block; margin: auto;" />


## Full vignette

You can see full vignette for model 8 at http://rpubs.com/cardiomoon/468602


## Shiny App

I have developed a shiny app. You can test the app at https://cardiomoon.shinyapps.io/processR.
I will appreciate any comment.

## How to perform this analysis with shiny app

You can see how to perform this analysis at http://rpubs.com/cardiomoon/468600

## Sample powerpoint file

In the shiny app, you can download the analysis results as a powerpoint file. You can download the sample file [model8.pptx](https://github.com/cardiomoon/processRDocs/blob/master/model8/model8.pptx?raw=true) - view with [office web viewer](https://view.officeapps.live.com/op/view.aspx?src=https://github.com/cardiomoon/processRDocs/blob/master/model8/model8.pptx?raw=true). 


NEWS.md
## R package processR version 0.2.8
=====================================
(2022-Apr-13)

- resubmission to CRAN


## R package processR version 0.2.7
=====================================
(2022-Jan-20)

- bug-fixed


## R package processR version 0.2.6
=====================================
(2021-Jan-5)

- Change shiny app URL

## R package processR version 0.2.5
=====================================
(2020-Mar-8)

- Make modelsSummaryTable function compatible with flextable(>= 0.5.8)


## R package processR version 0.2.4
===================================
(2020-Feb-22)

- Add 'drawCovar' argument to drawModel function 

- Can select 'estSE' as one of the 'whatLabel' argument of drawModel function


## R package processR version 0.2.3
===================================
(2020-Jan-25)

- Bug fixed

- CRAN release


## R package processR version 0.2.2.6
=====================================
(2019-Nov-18)

- independent on the package bda, jtools


## R package processR version 0.2.2.5
=====================================
(2019-Sep-3)

- Can use color palette with drawConcept and drawModel functions

## R package processR version 0.2.2.4
=====================================
(2019-Aug-6)

- Bug-fixed


## R package processR version 0.2.2.3
=====================================
(2019-July-27)

- Enhancement of drawModel, drawConcept and modelsSummary function

## R package processR version 0.2.2.2
=====================================
(2019-July-26)

- Bug-fixed


## R package processR version 0.2.2.1
=====================================
(2019-July-23)

- Enhancement of function drawModel and matrixPlot


## R package processR version 0.2.2.0
=====================================
(2019-July-18)

- Enhancement of function drawModel and drawConcept

- New functions label2name and setPositionNodes added

## R package processR version 0.2.1.9
=====================================
(2019-July-14)

- Enhancement of function multipleMediation

- Rename function drawModel2 to drawConcept


## R package processR version 0.2.1.8
=====================================
(2019-July-11)

- New function drawModel2 added

## R package processR version 0.2.1.7
=====================================
(2019-July-6)

- New function getBootData, densityPlot, qqPlot added 

## R package processR version 0.2.1.6
=====================================
(2019-June-30)

- New function matrixPlot, matrix2no added 


## R package processR version 0.2.1.5
=====================================
(2019-June-28)

- enhancement of drawModel

- multipleMediation can handle bmatrix

## R package processR version 0.2.1.4
=====================================
(2019-June-25)

- enhancement of drawModel


## R package processR version 0.2.1.3
=====================================
(2019-June-21)

- enhancement of multipleMediation

- can calculate specific indirect effect contrast in multiple mediator model

## R package processR version 0.2.1.2
=====================================
(2019-June-12)

- enhancement of drawModel and drawCatModel

- can calculate conditional effect of X on M, conditional effect of M on Y in conditional process model


## R package processR version 0.2.1.1
=====================================
(2019-June-12)

- enhancement of drawModel and drawCatModel

- can calculate index of moderated mediation in conditional process model with categorical X

## R package processR version 0.2.1.0
=====================================
(2019-June-10)

- statisticalDiagram and drawModel make plot with a list of lm object


## R package processR version 0.2.0.9
=====================================
(2019-June-9)

- New function extractIMM added 

## R package processR version 0.2.0.8
=====================================
(2019-June-5)

- add box.col argument to select fill color of box of pmacroModel, statisticalDiagram, drawModel and drawCatModel 


## R package processR version 0.2.0.7
=====================================
(2019-June-4)

- New function multipleMediation, makeCatEquation2 added


## R package processR version 0.2.0.6
=====================================
(2019-May-31)

- Bug fixed

## R package processR version 0.2.0.5
=====================================
(2019-May-29)

- Bug fixed

## R package processR version 0.2.0.4
=====================================
(2019-May-25)

- New function drawModel added

- Delete function drawCatModel2

- Rename function drawCatModel2 as drawCatModel


## R package processR version 0.2.0.3
=====================================
(2019-May-18)

- New function drawCatModel2 added


## R package processR version 0.2.0.2
=====================================
(2019-May-15)

- Enhancement of function catMediation, meanSummary and drawCatModel


## R package processR version 0.2.0.1
=====================================
(2019-May-11)

- New function print.modmedSummary2, modmedSummary2Table added


## R package processR version 0.2.0
=====================================
(2019-May-5)

- New function condPlotCat2, unfold added


## R package processR version 0.1.9.4
=====================================
(2019-May-3)

- New function adjustypos, drawCatModel added

## R package processR version 0.1.9.3
=====================================
(2019-May-1)

- New function makeCatModel, getCatSlopeDf, condPlotCat, makeCEDf, makeAnovaDf added

## R package processR version 0.1.9.2
=====================================
(2019-Apr-28)

- bug fixed

## R package processR version 0.1.9.1
=====================================
(2019-Apr-25)

- New function r2diff, modSummary3 and modSummary3Table added

## R package processR version 0.1.9
===================================
(2019-Apr-23)

- New function condPlot2 added


## R package processR version 0.1.8.2
=====================================
(2019-Apr-16)

- Bug fixed

- New function plotCoef, meanSummary, modSummary2, meanSummaryTable and modSummary2Table added

## R package processR version 0.1.8
===================================
(2019-Apr-9)


R/CorPlot.R
#' Make a table with correlation
#'
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @export
corTable=function(fit){
    data=fit@Data@X[[1]]
    colnames(data)=fit@Data@ov$name
    data=data.frame(data)
    data
    result=mycor(data)
    result
    resp=p2asterisk(result$p)
    res=paste0(sprintf("%.2f",result$r),resp)
    res=matrix(res,ncol(result$r))
    res[upper.tri(res)]=""
    res[row(res)==col(res)]<-"1"
    rownames(res)=colnames(data)
    colnames(res)=colnames(data)
    res=as.data.frame(res)
    # str(res)
    res
    values=unique(as.vector(resp))
    temp=""
    if("***" %in% values) temp=pastecolon(temp,"***p<0.001")
    if("**" %in% values) temp=pastecolon(temp,"**p<0.01")
    if("*" %in% values) temp=pastecolon(temp,"*p<0.05")
    attr(res,"footer")=temp
    res
}

#' Make a table with correlation
#'
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @param vanilla Logical. If true, vanilla.table is returned
#' @param addFooter Logical. If true, footer added
#' @param seek string to look for
#' @param replace A string of replacement
#' @importFrom rrtable df2flextable
#' @importFrom flextable align color add_footer merge_at
#' @export
corTable2=function(fit,vanilla=TRUE,addFooter=FALSE,seek=NULL,replace=NULL){

    res=corTable(fit)
    if(!is.null(seek)){
        for(i in seq_along(seek)) {
    colnames(res)[colnames(res)==seek[i]]=replace[i]
    rownames(res)[rownames(res)==seek[i]]=replace[i]
        }
    }
    Table=rrtable::df2flextable(res,vanilla=vanilla,add.rownames=TRUE)
    Table=flextable::align(Table,align="center",part="all")
    Table=flextable::color(Table,i=1,j=1,color=ifelse(vanilla,"white","#5B7778"),part="header")

    if(addFooter){
        Table<-flextable::add_footer(Table,rowname=attr(res,"footer"))
        Table<-flextable::merge_at(Table,j=1:(ncol(res)+1),part="footer")
        Table=flextable::align(Table,align="right",part="footer")
    }
    Table
}


#' paste two character with colon
#' @param temp a character
#' @param x a character
pastecolon=function(temp,x){
    if(temp=="") res=x
    else res=paste0(temp," ; ",x)
    res
}

#' Convert p values to asterisk
#' @param x a numeric vector or matrix
p2asterisk=function(x){
    ifelse(x<0.001,"***",ifelse(x<0.01,"**",ifelse(x<0.05,"*","")))
}



#'Extract model fit measures to data.frame
#'
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @param digits integer indicating the number of decimal places to be used.
#' @param names names of statistic to be extracted
#'
#' @importFrom lavaan fitMeasures
#' @export
#' @return A data.frame
modelFitTable=function(fit,digits=2,names=NULL){
    # digits=2;names=NULL
    if(is.null(names)) {
        names=c("chisq","df","pvalue","cfi","gfi","agfi","tli","rmr","srmr","rmsea","rmsea.ci.lower","rmsea.ci.upper","aic","bic")
        newnames=c("chisq","df","p","CFI","GFI","AGFI","TLI","RMR","SRMR","RMSEA","lower","upper","AIC","BIC")

    }
    names
    value=fitMeasures(fit)[names]
    value
    res=data.frame(rbind(value))
    if(!is.null(newnames)) colnames(res)=newnames
    rownames(res)="statistic"
    res$x2df=res$chisq/res$df
    res=round(res,digits)
    res$RMSEA=paste0(res$RMSEA,"(",res$lower,"-",res$upper,")")
    colnames(res)[10]="RMSEA(95% CI)"
    res=res[c(1,2,15,3:10,13,14)]
    res
}


#'Extract model fit measures to flextable
#'@param fit An object of class lavaan. Result of sem function of package lavaan
#'@param vanilla Logical
#'@param ... Further arguments to be passed to modelFitTable()
#'@export
modelFitTable2=function(fit,vanilla=FALSE,...){
    result=modelFitTable(fit,...)
    df2flextable(result,vanilla=vanilla)
}


#'convert parameterEstimates to data.frame
#'
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @param latent whether the latent variables be included in result
#' @param regression whether the regressions be included in result
#' @param mediation whether the mediation effects be included in result
#' @param covar whether the covariances be included in result
#' @param ci If TRUE, confidence intervals are added to the output
#' @param standardized Logical. If TRUE, standardized estimates are added to the output
#' @param digits integer indicating the number of decimal places to be used.
#'
#' @export
estimatesTable=function(fit,latent=TRUE,regression=TRUE,mediation=FALSE,covar=FALSE,ci=FALSE,standardized=TRUE,digits=2){
     # latent=TRUE;regression=TRUE;mediation=FALSE;covar=FALSE;ci=TRUE;standardized=TRUE;digits=2
    # cat("digits=",digits,"\n")
    result=parameterEstimates(fit,ci=ci,standardized=standardized)
    result
    if(mediation){
       result=result[result$label!="",]
       result=result[-c(2)]
    } else{
        include=c()
        if(latent) include=c(include,"=~")
        if(regression) include=c(include,"~")
        if(covar) include=c(include,"~~")
        result=result[result$op %in% include,]
        #result=result[-c(2,4)]
        result=result[-c(2)]
    }
    no=ncol(result)
    result=result[-c(no-2,no)]
    result
    if(ci){
        result$est=paste0(round(result$est,digits),"(",round(result$ci.lower,digits),"-",round(result$ci.upper,digits),")")
        no=ncol(result)
        result=result[-c(no-1,no-2)]
    } else{
        result$est=round(result$est,digits)
    }
    result
    result$se=round(result$se,digits)
    result$z=round(result$z,digits)
    result$std.all=round(result$std.all,digits)
    result$pvalue=convertPvalue(result$pvalue)
    result
    if(mediation){
        result$lhs[substr(result$lhs,1,3)=="ind"]="indirect effect"
        result$lhs[substr(result$lhs,1,5)=="total"]="total effect"
        colnames(result)=c("Variables","Predictors","label","B","SE","z","p","\u03B2")

    } else{
        colnames(result)=c("Variables","Predictors","label","B","SE","z","p","\u03B2")
    }
    result[is.na(result)]=""
    result
}


#'convert parameterEstimates to flextable
#'@param fit An object of class lavaan. Result of sem function of package lavaan
#'@param vanilla Logical
#'@param digits integer indicating the number of decimal places to be used.
#' @param seek string to look for
#' @param replace A string of replacement
#'@param ... Further arguments to be passed to estimatesTable()
#'@export
estimatesTable2=function(fit,vanilla=TRUE,digits=3,seek=NULL,replace=NULL,...){
    result=estimatesTable(fit,digits=digits,...)
    if(!is.null(seek)){
        for(i in seq_along(seek)){
        result$Predictors[result$Predictors==seek[i]]=replace[i]
        }
    }
    df2flextable(result,vanilla=vanilla,digits=digits)
}


#' convert vector of p values to string
#'
#' @param x vector of p values
convertPvalue=function(x){
    x=sprintf("%0.3f",x)
    x[x=="NA"]=""
    x[x=="0.000"]="< 0.001"
    x
}



#' Draw correlation plot
#'
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @param label if 0, no label(default), if 1, use r value as label, if 2, use r value with significant mark as label
#' @param yreverse Logical. if true, reverse the order of y axis.
#' @param xangle axis.x.text.angle
#' @param seek string to look for
#' @param replace A string of replacement
#' @param ... Further arguments to be passed on to geom_text
#'
#' @export
#'
#' @return A ggplot
corPlot=function(fit,label=2,yreverse=TRUE,xangle=45,seek=NULL,replace=NULL,...){
    data=fit@Data@X[[1]]
    colnames(data)=fit@Data@ov$name
    data=data.frame(data)
    data
    if(!is.null(seek)) {
        for(i in seq_along(seek)) {
        colnames(data)[colnames(data)==seek[i]]=replace[i]
        }
    }
    ggCor(data=data,label=label,yreverse=yreverse,xangle=xangle)
}


#' Make a Cronbach alpha table
#' @param fit An object of class lavaan. Result of sem function of package lavaan
#' @param digits integer indicating the number of decimal places to be used.
#'
#' @importFrom psych alpha
#' @importFrom stringr str_flatten
#' @export
fit2alpha=function(fit,digits=3){
        df=as.data.frame(fit@ParTable,stringsAsFactors = FALSE)
        df=df[df$op=="=~",c("lhs","rhs")]

        latentVars=unique(df$lhs)
        indicators=c()
        data=fit@Data@X[[1]]
        fit@Data@X[[1]]
        colnames(data)=fit@Data@ov$name
        data=data.frame(data)
        data
        alpha<-lambda6<-c()

        for(i in seq_along(latentVars)){
            vars=df$rhs[df$lhs==latentVars[i]]
            indicators=c(indicators,stringr::str_flatten(vars,"+"))
            data[vars]
            result=psych::alpha(data[vars],warnings=FALSE)
            alpha=c(alpha,result$total$raw_alpha)
            lambda6=c(lambda6,result$total$`G6(smc)`)
        }
        alpha=round(alpha,digits)
        lambda6=round(lambda6,digits)
        data.frame(latentVars,indicators,alpha,lambda6,stringsAsFactors = FALSE)
}

